app/src/org/commcare/android/tasks/DataPullTask.java

```diff
@@ -219,196 +232,67 @@ public void reportProgress(int progress) {
                     purgeCases();
                     useRequestFlags = true;
                 }
-                //Either way, don't re-do this step
-                this.publishProgress(PROGRESS_CLEANED);
-
-                int responseCode = -1;
-                HttpResponse response = null;
                 
-                //This isn't awesome, but it's hard to work this in in a cleaner way
-                if(DEBUG_LOAD_FROM_LOCAL) {
-                    mDebugStream = this.c.getAssets().open("payload.xml");
-                    responseCode = 200;
-                } else {
-                    response = requestor.makeCaseFetchRequest(server, useRequestFlags);
-                    responseCode = response.getStatusLine().getStatusCode();
-                }
-                Logger.log(AndroidLogger.TYPE_USER, "Request opened. Response code: " + responseCode);
+                boolean isMoreData = true;
                 
-                if(responseCode == 401) {
-                    //If we logged in, we need to drop those credentials
-                    if(loginNeeded) {
-                        CommCareApplication._().logout();
+                //Make sure this is primed
+                this.mRetryAt = -1;
+                //Ok, so it's time to start trying the sync. This is in a loop in case the server requests that
+                //we need to wait for further data
+                while(isMoreData) {
+                    if(this.isCancelled()) {
+                        return RESULT_CANCELLED;
                     }
-                    Logger.log(AndroidLogger.TYPE_USER, "Bad Auth Request for user!|" + username);
-                    return AUTH_FAILED;
-                } else if(responseCode >= 200 && responseCode < 300) {
-                    
-                    if(loginNeeded) {                        
-                        //This is necessary (currently) to make sure that data
-                        //is encoded. Probably a better way to do this.
-                        CommCareApplication._().logIn(CryptUtil.unWrapKey(ukr.getEncryptedKey(), password), ukr);
-                        wasKeyLoggedIn = true;
+                    //Check whether we're currently waiting for a retry timer to expire
+                    if(mRetryAt != -1 && !(System.currentTimeMillis() > mRetryAt)) {
+                        //if we are, wait for a short amount of time and then jump back to the beginning of the loop.
+                        //we have to poll this to be able to check for user cancel inputs.
+                        try {
+                            int secondsUntilSync = (int)(mRetryAt - System.currentTimeMillis());
+                            int secondsSinceStart = (int)(System.currentTimeMillis() - this.mRetryBegan);
+                            this.publishProgress(PROGRESS_SERVER_PROCESSING, secondsUntilSync, secondsSinceStart);

```

Comment:
These are milliseconds, how about calling them millisUntilSync and millisSinceStart?